### PR TITLE
test/fix: add tests for, and address bugs with i32 icmp

### PR DESCRIPTION
**Pull request by bitwalker**
NOTE: I'm still adding more tests, looking for more bugs, so this shouldn't be merged yet, but can be reviewed. As I add to it, I'll do so in new commits to aid in re-review.

### CODE REVIEW

**Comment by bitwalker** [comment]
@greenhat This PR will contain any fixes/tests for the i32 intrinsics. It's based on the branch from #37, so that will need to be merged first.

File: codegen/masm/intrinsics/i32.masm
Comment (greenhat) [inline codegen/masm/intrinsics/i32.masm:25]:
```suggestion
    dup.1 eq # [is_same_sign, is_b_signed, b, a]
```

File: codegen/masm/intrinsics/i32.masm
Comment (greenhat) [inline codegen/masm/intrinsics/i32.masm:31]:
```suggestion
    dup.0 exec.is_signed # [is_result_signed, result, is_same_sign, is_b_signed]
```

File: codegen/masm/src/tests.rs
Comment (greenhat) [inline codegen/masm/src/tests.rs:457]: As a side note, `proptest` should be a good fit for these kinds of tests. Its strategy for integer types covers such edge cases.

**Review by greenhat — APPROVED** [review]
LGTM!

File: codegen/masm/intrinsics/i32.masm
Comment (bitwalker) [reply codegen/masm/intrinsics/i32.masm:25]: I wanted to make it clear that the purpose of that value after computing `is_same_sign` is to indicate whether `a` and `b` are signed (when `is_same_sign` is true). When `is_same_sign` is false, it's irrelevant whether `is_signed` refers to `a` or `b` specifically, since when the signs differ, overflow is impossible.

That said, if you found it confusing, I'm happy to commit this suggestion. What do you think?

File: codegen/masm/src/tests.rs
Comment (bitwalker) [reply codegen/masm/src/tests.rs:457]: For sure, I mainly wanted to get a few sanity tests implemented so I can unblock you on these intrinsics, but I plan to revisit many of these tests to do property-based testing :)

File: codegen/masm/intrinsics/i32.masm
Comment (greenhat) [reply codegen/masm/intrinsics/i32.masm:25]: Well, after the explanation, it makes sense. Although it still requires more cognitive overhead, I'm ok with it.

Comment (bitwalker) [reply codegen/masm/intrinsics/i32.masm:25]: I'll rewrite the comments here to make it clearer upon first reading - I'd rather not risk it be confusing.

**Comment by bitwalker** [comment]
@greenhat This PR is ready to merge. I've rewritten the intrinsics tests with `proptest`, as you suggested, and we now have coverage for the entire `intrinsics::i32` module tests-wise (a few functions which are basically simple wrappers are not explicitly tested, but their interesting behavior _is_ tested, e.g. `wrapping_add` and `checked_add` are wrappers for `overflowing_add` which do something appropriate with the `overflowed` flag). 

Once we get this merged, we should be good to go to start working with Wasm modules that have signed `i32` ops!

**Pull request merged**
